### ISSUE TITLE
[BUG] IchimokuChikouSpanIndicator#calculate returns close price in the future :)

### ISSUE THREAD

**Issue by jaric85** [ta4j/ta4j#614]
According to https://school.stockcharts.com/doku.php?id=technical_indicators:ichimoku_cloud 

> Chikou Span (Lagging Span): Close plotted 26 days in the past
> The default setting is 26 periods, but can be adjusted

IchimokuChikouSpanIndicator should return close price in the **past** not the **future**, am I missing something here?

**To Reproduce**
1. Create any strategy relying only on IchimokuChikouSpanIndicator
2. strategy.shouldEnter(index), if index > lastIndexOnBarSeries - (IchimokuChikouSpanIndicator#timeDelay - 1)  indicator would always return NaN.NaN.

**Expected behavior**
```
    @Override
    protected Num calculate(int index) {
        if(index - timeDelay < 0 ){
            return NaN.NaN;
        }
        return closePriceIndicator.getValue(index - timeDelay);
    }
```

**Comment by team172011** [comment]
Hi @jaric85 I do not understand where you see a bug here. 

It returns the value of the close price indicator in the past (timeDelay). Only if your `timeDelay `is bigger than the `index `you are requesting, it returns NaN because you do not have any information about close prices before the start of your bar series .

**Comment by jaric85** [comment]
Hi,
Please take a look at the current implementation on master HEAD and let me know if you are still sure about your statement, that it returns close price in the past. Code snippet above is expected implementation and current implementation is
`  @Override
    protected Num calculate(int index) {
        int spanIndex = index + timeDelay;
        if (spanIndex <= getBarSeries().getEndIndex()) {
            return closePriceIndicator.getValue(spanIndex);
        } else {
            return NaN.NaN;
        }
    }  `
 
It doesn't make any sense, to me, this part

int spanIndex = index + timeDelay;

Do you agree that substraction should be used instead?

Regards,
Goran

**Comment by team172011** [comment]
Hi, I see you are right. 
I thought your code snipped was from the master HEAD, but It is a suggestion how to fix. Yes the current implementation makes is absolutely senseless and we should it fix with your proposal

**Comment by team172011** [comment]
also the junit tests are strange :D

**Comment by MarcDahlem** [comment]
Yesterday I struggeled with the same thing as some tests failed after a rework. 
 would guess that the IchimokuChikouSpanIndicator  and its test use weird behaviours of our caching.

Especially this line in the CachingIndicator causes errors (or fixes, whatever) the Ichinmoku Tests: https://github.com/ta4j/ta4j/blob/master/ta4j-core/src/main/java/org/ta4j/core/indicators/CachedIndicator.java#L107

**Comment by team172011** [comment]
not fixed - continued in #705
